d2/src/test/java/com/linkedin/d2/balancer/subsetting/DeterministicSubsettingStrategyTest.java

```diff
@@ -0,0 +1,204 @@
+          .thenReturn(new DeterministicSubsettingMetadata(i, clientNum));
+      _deterministicSubsettingStrategy = new DeterministicSubsettingStrategy<>(_deterministicSubsettingMetadataProvider,
+          "test", minSubsetSize);
+      Map<String, Double> weightedSubset = _deterministicSubsettingStrategy.getWeightedSubset(pointsMap, 0);
+      assertTrue(weightedSubset.size() >= Math.min(minSubsetSize, hostNum));
+
+      for (Map.Entry<String, Double> entry: weightedSubset.entrySet())
+      {
+        distributionMap.put(entry.getKey(), distributionMap.getOrDefault(entry.getKey(), 0D) + entry.getValue());
+      }
+    }
+
+    double host0WeightSum = distributionMap.getOrDefault("host0", 0D);
+    for (double weightSum: distributionMap.values())
+    {
+      assertEquals(weightSum, host0WeightSum, DELTA_DIFF);
+    }
+  }
+
+  @Test(dataProvider = "differentWeightsData")
+  public void testDistributionWithDifferentWeights(int clientNum, double[] weights, int minSubsetSize)
+  {
+    Map<String, Double> pointsMap = constructPointsMap(weights);
+    Map<String, Double> distributionMap = new HashMap<>();
+
+    for (int i = 0; i < clientNum; i++)
+    {
+      Mockito.when(_deterministicSubsettingMetadataProvider.getSubsettingMetadata())
+          .thenReturn(new DeterministicSubsettingMetadata(i, clientNum));
+      _deterministicSubsettingStrategy = new DeterministicSubsettingStrategy<>(_deterministicSubsettingMetadataProvider,
+          "test", minSubsetSize);
+      Map<String, Double> weightedSubset = _deterministicSubsettingStrategy.getWeightedSubset(pointsMap, 0);
+
+      for (Map.Entry<String, Double> entry: weightedSubset.entrySet())
+      {
+        distributionMap.put(entry.getKey(),
+            distributionMap.getOrDefault(entry.getKey(), 0D) + entry.getValue() * pointsMap.get(entry.getKey()));
+      }
+    }
+
+    double totalWeights = distributionMap.values().stream().mapToDouble(Double::doubleValue).sum();
+    double totalHostWeights = Arrays.stream(weights).sum();
+    for (Map.Entry<String, Double> entry: distributionMap.entrySet())
+    {
+      String hostName = entry.getKey();
+      double hostWeight = weights[Integer.parseInt(hostName.substring("test".length()))];
+      assertEquals(entry.getValue() / totalWeights, hostWeight / totalHostWeights, DELTA_DIFF);
+    }
+  }
+
+  @DataProvider
+  public Object[][] uniformWeightData()
+  {
+    return new Object[][]
+        {
+            {5, 0, 10},
+            {1, 1, 10},
+            {1, 5, 10},
+            {5, 1, 10},
+            {5, 5, 10},
+            {5, 5, 1},
+            {3, 6, 2},
+            {5, 40, 10},
+            {10, 100, 10},
+            {7, 47, 13},
+            {47, 40, 13},
+            {13, 200, 11},
+            {83, 359, 23}
+        };
+  }
+
+  private static double[] generateRandomWeights(int size)
+  {
+    return new Random().doubles(size, 0D, 1D).toArray();
```

Comment:
I haven't looked into the detailed calculation implementation yet, just want to double check that we're not making assumption that the weight is always in the scale of 0-1. 
I think Dyno might increate the weight to something greater than 1.

Reply:
Yes, it will work with weight in all scales.

Reply:
Probably worth adding some test cases to verify this